File: src/leetcode/p0100_0199/p0100_same_tree/SameTree.java
/*
 * Copyright (C) 2021  Weiyang(Will) Xu
 * 
 * This file is part of algorithm.
 * Algorithm is free software: you can redistribute it and/or modify
 * it under the term of the GNU General Public License version 3 or
 * any later version, as specified in the readme.md file.
 */

package leetcode.p0100_0199.p0100_same_tree;

import common.provide.TreeNode;

public interface SameTree {
    /**
     * Given the roots of two binary trees p and q, write a function to
     * check if they are the same or not.
     *
     * Two binary trees are considered the same if they are
     * structurally identical, and the nodes have the same value.
     *
     * Constraints:
     * - The number of nodes in both trees is in the range [0, 100],
     * - -10^4 <= Node.val <= 10^4
     *
     * @param p 0 <= p.size <= 100
     *          -10^4 <= Node.val <= 10^4
     * @param q 0 <= q.size <= 100
     *          -10^4 <= Node.val <= 10^4
     * @return same or not
     */
    boolean isSameTree(TreeNode p, TreeNode q);
}

File: src/leetcode/p0100_0199/p0100_same_tree/same_tree.py
from abc import ABC, abstractmethod
from typing import Optional

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class SameTree(ABC):
    """
    Given the roots of two binary trees p and q, write a function to check if they are the same or not.

    Two binary trees are considered the same if they are structurally identical, and the nodes have the same value.
    """
    @abstractmethod
    def is_same_tree(self, p: Optional[TreeNode], q: Optional[TreeNode]) -> bool:
        """
        Args:
            p, q:
                - The number of nodes in both trees is in the range [0, 100].
                - -10^4 <= Node.val <= 10^4
        """
        pass

class SameTreeRecursive(SameTree):
    def is_same_tree(self, p: Optional[TreeNode], q: Optional[TreeNode]) -> bool:
        if not p and not q:
            return True
        if not p or not q or p.val != q.val:
            return False

        left_same = self.is_same_tree(p.left, q.left)
        right_same = self.is_same_tree(p.right, q.right)
        return left_same and right_same
